### ISSUE TITLE
2024-12-16 - KDB Project Call Meeting Minutes

### ISSUE THREAD

**Issue by alvin-c-shih** [finos/kdb#114]
## Date
Monday, December 16, 2024 - 9:30am ET / 2:30pm UK

_// Second Monday of every month_

## Attendees
| Fullname | Affiliation | GitHub Username |
|:-----|:-----|:-----|
| Alvin Shih            | Morgan Stanley    | @alvin-c-shih      
| Rian Ó Cuinneagáin    | KX Systems        | @rianoc            
| Peter Gyorok          | KX Systems        | @gyorokpeter       
| Ross Duffy            | Morgan Stanley    | @rdffy             
| Aaron Davies          | Morgan Stanley    | @adavies42         
| Karl Moll             | FINOS             | @karlmoll   
| Adam Honore | AWS | 

## Meeting notices
- FINOS **Project leads** are responsible for observing the FINOS guidelines for [running project meetings](https://github.com/finos/community/blob/master/governance/Meeting-Procedures.md#run-the-meeting). Project maintainers can find additional resources in the [FINOS Maintainers Cheatsheet](https://odp.finos.org/docs/finos-maintainers-cheatsheet/).

- **All participants** in FINOS project meetings are subject to the [LF Antitrust Policy](https://www.linuxfoundation.org/antitrust-policy/), the [FINOS Community Code of Conduct](https://github.com/finos/community/blob/master/governance/Code-of-Conduct.md) and all other [FINOS policies](https://github.com/finos/community/tree/master/governance#policies). 

- FINOS meetings involve participation by industry competitors, and it is the intention of FINOS and the Linux Foundation to conduct all of its activities in accordance with applicable antitrust and competition laws. It is therefore extremely important that attendees adhere to meeting agendas, and be aware of, and not participate in, any activities that are prohibited under applicable US state, federal or foreign antitrust and competition laws. Please contact legal@finos.org with any questions.

- FINOS project meetings may be recorded for use solely by the FINOS team for administration purposes. In very limited instances, and with explicit approval, recordings may be made more widely available.

## Agenda

- (5 min) Convene, roll call, welcome new people.
- [ ] Display [FINOS Antitrust Policy summary slide](https://github.com/finos/community/blob/master/governance/Compliance-Slides/Antitrust-Compliance-Slide.pdf) 
- [ ] Review Meeting Notices (see above)
- (5 min) Approve previous meeting minutes.
  - Approved by acclamation.
- (5 min) Got a message from Neil Kanungo on LinkedIn asking to populate a survey on AI-assisted programming kdb+.
  - Will be very interesting to see if KX can get an LLM to understand `q` code better than what's currently out there!
  - Listed various IDEs / editors, but Emacs was not among them.
  - Most open source integrations with GitHub Copilot just use the "language server".
    - https://stackoverflow.com/questions/76741410/how-to-invoke-github-copilot-programmatically
    - That only supports "completions" and not "chat".
    - Completions model is GPT-3.5 Turbo.
    - Not as advanced as the chat model, GPT-4o. 
- (10 min) qStudio is now free?!
  - https://www.timestored.com/b/qstudio-is-now-free/
  - https://github.com/timestored/qstudio
  - Would it be better to follow Jo Shinonome's lead and drink the VS Code Kool-Aid?
    - https://github.com/jshinonome/vscode-q
    - KX has a VS Code extension as well.
      - https://marketplace.visualstudio.com/items?itemName=KX.kdb
      - Might have limited uptake at enterprise customers that require custom authentication.
    - kdb Studio still quite heavily used for data exploration.
      - https://github.com/finos/kdb-studio
    - VS Code more popular for code development.
  - AWS Q Developer may work on some kdb/q questions.
    - Fully supported languages listed on FAQ page.
      - https://aws.amazon.com/q/developer/faq
  - Jo Shinonome's language server connects to a `q` process to see functions and namespaces.
    - Can hover over a symbol to get a definition.
    - KX extension has a hover over as well.
  - AI-powered tools still require the developer to iterate on ideas.
    - [The 70% problem: Hard truths about AI-assisted coding](https://addyo.substack.com/p/the-70-problem-hard-truths-about)
- (10 min) PyKX 3.0 Officially Released
  - Might be interesting to try the free tier of FinSpace to compare what AWS vs tickerplant-launching classes of PyKX.
    - https://aws.amazon.com/finspace/features/managed-kdb-insights/
    - https://code.kx.com/pykx/3.0/api/tick.html#pykx.tick.BASIC
- (5 min) Ryan Hamilton taking issue with PyKX being branded as "open source".
  - https://www.linkedin.com/posts/justryanhamilton_pykx-is-closed-source-what-do-you-think-activity-7272927475511259137-dzNl
  - https://github.com/KxSystems/pykx
    - `This work is dual licensed under "Apache 2.0" and the "Software License for q.so", users are required to abide by the terms of both licenses in their entirety.`
  - Naive interpretations of "open source" are disingenuous given the recognition of organizations like OSI, Linux Foundation, etc.
    - https://opensource.org/
    - https://www.linuxfoundation.org/
  - Just call it "source available" so that users can adapt the code to their local setups.
  - Many `.q_` files in the repo aren't "source code".
    - Should separate these into their own repo with a different license.
- (10 min) qPython updated for `numpy/1.24` based on suggestion from AI assistant.
  - https://github.com/finos/qPython/pull/22
  - Should be straightforward search/replace after release of that to update for `numpy/2.0`.
  - Agentic AI SWE went down a bit of a rabbit hole...
- (5 min) Broken link to "KX Whitepaper On Using IPC With kdb+"?
  - https://kx.com/blog/kx-whitepaper-on-using-ipc-with-kdb/
  - Available on Archive:
    - https://web.archive.org/web/20210413162939/https://code.kx.com/q/wp/ipc/
- (5 min) AOB, Q&A, Adjourn.
  - Typo in https://github.com/finos/kdb/issues/86 .
    - `disambuguate` -> `disambiguate`
  - PyKX words might clash with user variable names.
  - Would have to be conditional / explicitly activated.
  - Unfortunately, only `.q` namespace allows infix.
    - Would be useful to allow "blessing" of other namespaces.
  - Might be interesting to use AI Coding Assistant to learn how Shakti parser works.

## Join Zoom Meeting
- https://zoom.us/j/94523415326?pwd=NkJDaUtUUlZTcHVwcUJHdG0yRFJUdz09
- Meeting ID: 945 2341 5326
- Passcode: 361990
- Find your local number: https://zoom.us/u/abTtYRZ0B6

**Comment by karlmoll** [comment]
Karl Moll / FINOS

**Comment by alvin-c-shih** [comment]
Alvin Shih / Morgan Stanley

**Comment by gyorokpeter** [comment]
Péter Györök / KX

**Comment by adavies42** [comment]
Aaron Davies/Morgan Stanley

**Comment by rdffy** [comment]
Ross Duffy / Morgan Stanley

**Comment by rianoc-kx** [comment]
Rian Ó Cuinneagáin / KX

**Comment by rdffy** [comment]
pykx 3.0 providing a solution to this:
https://github.com/finos/kdb/issues/86

**Comment by rianoc-kx** [comment]
Adam's link: https://aws.amazon.com/finspace/features/managed-kdb-insights/

**Comment by karlmoll** [comment]
https://communications.dtcc.com/dtcc-hackathon-registration-lp-16471.html

**Comment by adavies42** [comment]
https://web.archive.org/web/20210331173202/https://kx.com/blog/kx-whitepaper-on-using-ipc-with-kdb/

**Comment by rdffy** [comment]
https://github.com/jshinonome/vscode-q Jo's vscode-q
https://marketplace.visualstudio.com/items?itemName=KX.kdb KX's vscode extension

Both have a q language server as part of their extensions so that you get some autocompletion and linting capabilities.
As well as being able to jump to function definitions and things like that

**Comment by jshinonome** [comment]
> https://github.com/jshinonome/vscode-q Jo's vscode-q
> 
> https://marketplace.visualstudio.com/items?itemName=KX.kdb KX's vscode extension
> 
> 
> 
> Both have a q language server as part of their extensions so that you get some autocompletion and linting capabilities.
> 
> As well as being able to jump to function definitions and things like that 
> 
> 
> 
> 

Indeed, my language server possesses superior capabilities in analyzing actual code and generating a syntax tree. The advanced language server (provided by q Professional) also offers the ability to format and lint q code. However, Kx lacks the expertise to construct a language server effectively. All programming languages have dedicated teams responsible for maintaining language servers, as this is the only viable method for ensuring the accessibility of programming languages to new developers. It is inappropriate to compare my language server to Kx’s. They operate at significantly different levels of sophistication. Furthermore, the Kx actually copied my language server binary in the initial release without authorization. However, it was only removed after I made a request. Making open-source projects for kdb sounds ridiculous.
